### ISSUE TITLE
A doubt about the gbin.md format

### ISSUE THREAD

**Issue by Eldinen** [u3shit/neptools#8]
I am trying to figure out the formats of cl3 and gbin to be able to do changes in the code. The cl3 format is almost clear, thanks to your docs, but the gbin... well I dont get the struct and string table format to be able to recalculate the pointers.

Could you extend the doc formats information?
Maybe my problem is I am reading your docs and looking in these files.
https://drive.google.com/open?id=0B2552rsEmf9wbWFZVWV3LV82R1k

**Comment by u3shit** [comment]
Hm, the files are cl3 files, you have to extract the gbin files (neptools can do it). The gbin format is a bit more complicated, as the file itself describes the format of the data stored (even though the game engine mostly ignores it). The type descriptor describes the format, check check it before trying to understand the rest of the file.

**Comment by Eldinen** [comment]
the gbin from database are easy, but the one from main.CL3 is as you say more complicated. But at the end, I only need the information with the type 5 from the type descriptor, right?

**Comment by u3shit** [comment]
main.cl3 is a bit complicated... It contains a main.dat, which is like a bytecode for some weird VM used by the game, and the gbin file is just embedded into that. And if you want to change the length of the gbin, you'll have to recalculate all offsets. There's no documentation for the .dat (stcm) format, because even neptools doesn't understand that files completely.

**Comment by Eldinen** [comment]
Yes, no problem about stcm format, I think I have got it hehe, 
yes... gbin/stcm will be longer so I need to recalculate all offsets in the cl3, but I dont find a way to know where they are. So to be able to recalculate all offsets, could you tell me more about main.cl3?

**Comment by u3shit** [comment]
main.cl3 is just a normal cl3 file, see `doc/fomats/cl3.md`

**Comment by Eldinen** [comment]
But the offsets under GBNL header inside main.cl3 i dont understand how they must be calculated and even where they are

**Comment by u3shit** [comment]
The offsets under the GBNL header are relative to the beginning of the gbnl file. Since the GBNL is embedded inside an STCM file, you'll have to figure out where the GBNL starts from the STCM.

**Comment by Eldinen** [comment]
Yes I know where the GBNL starts, but I dont know how to find the offsets under GBNL footer that I have to calculate when GBNL gets longer

**Comment by Eldinen** [comment]
How did you find the offsets under GBNL?

**Comment by u3shit** [comment]
What kind of offsets?

**Comment by Eldinen** [comment]
As you said, if I want to change the length of the gbin inside main.cl3, i'll have to recalculate all offsets of the cl3. I saw your tool do this for your neps cl3 files, so how did you find where are the offsets that you have to recalculate?

**Comment by u3shit** [comment]
Oh, that one? Well, that's a bit complicated, since you have to parse the whole stcm stuff and I haven't written a documentation for that. Your best bet is looking into `src/format/stcm` for the time being.

**Comment by Eldinen** [comment]
What file in that folder is the one which recalculate the offsets?

**Comment by u3shit** [comment]
It's not that simple. It parses the file into a lot of structures (and there are no offsets here, only pointers), then serializes it back again on saving.
I think I should put writing an stcm format documentation on my todo list...

**Comment by Eldinen** [comment]
That would be really helpful!
Yes, what I want to know is how to find the pointers I have to recalculate

**Comment by u3shit** [comment]
https://github.com/u3shit/neptools/blob/master/doc/formats/stcm.md
Comment if something is not clear, it's not a simple file.

**Comment by Eldinen** [comment]
Thank you! I will study it when i get home. I am out of my country now.

Really thank you!

**Comment by Eldinen** [comment]
I have had time to study the STCM file, and I am a bit confuse with the Instruction struct but more confuse with its Parameters and types. Dont undesrtand how to read them to be able to recalculate offsets.

Besides I see for example 3 instructions headers after reading Export entry offset, but there is not any place where I can know how many instructions there are for that export entry (I am talking about main export entry name)

**Comment by u3shit** [comment]
No, the format doesn't store how many instructions are there, nor their positions other than the few exported ones. Neptools does a simple control flow analysis to find out the instructions.

I don't know which Neptunia game you're seeing. In r;b1's `0001/main.cl3`'s `main.DAT` at 0x90 there is
```
00000090  00 00 00 00 01 00 00 00  01 00 00 00 e0 03 00 00  |................|
000000a0  e9 b7 c4 04 0c 00 00 00  ff ff ff ff ff ff ff ff  |................|
```
so `type=0, offset_unit=1, field_8=1, length=0x3e0`, and the gbnl data follows.

When `is_call==1`, the instruction is a call instruction, and the vm will jump to the instruction at `opcode_offset`, so yeah, you should care about it. (If `is_call==0`, it's an opcode, so that's when you should ignore it.)

**Comment by Eldinen** [comment]
That is what I thought, ok, lets do a simple control hehe.
The tests I am doing are with Neptunia rb3, V generation.

Ups, yes I saw it, forgive me,  Its exactly as your tools say.

Then, if is_call==1, in what way should I care about it?

**Comment by u3shit** [comment]
>Then, if is_call==1, in what way should I care about it? I mean, what do I have to do with the place where that offset is pointing?

There will be an another set of instructions there.

**Comment by Eldinen** [comment]
Really true! Thanks your documentation is absolutly perfect, maybe I should read it only when I am nos sleepy :) thanks again. I think I will close this issue soon.
